### DISCUSSION TITLE
MVVM 적용하여 리펙토링중입니다.

### DISCUSSION THREAD

**Discussion by SSSiye** [DeveloperAcademy-POSTECH/.github#104] [Q&A]
아키텍처라는것을 모른체로 매크로를 끝내고 앱 배포를 완료한 상태입니다.

이제서야 업데이트를 위해 리펙토링을 합니다.

--------------------------------------------------------
<img width="584" alt="image" src="https://github.com/user-attachments/assets/8182659d-6bc3-4118-96a3-09ff4477411f" />

--------------------------------------------------------
<img width="641" alt="image" src="https://github.com/user-attachments/assets/0cc0ef5a-c816-4444-bc6e-c0cff78a1ce0" />

--------------------------------------------------------
현재는 이런 상태이구요. (코드 구조가 굉장히 엉망진창😢)
한 View당 하나의 Viewmodel을 만들어야 한다고 알고있습니다. 그렇다면 제가 사용한 ShoppibngViewModel,ListViewModel과 같은 '여러 뷰에서 같은 데이터 값을 사용하는 것'들도 각각의 ViewModel을 만들어서 관리해야하나요?

**Comment by moonkey48** [comment]
헤이 씨씨! 일단 리팩토링을 시작하시게 된 것 너무 축하드립니다. 

각각의 ViewModel을 만들어야 하는지에 대한 답변 이전에, 아래 질문에 대한 고민을 해보시는 것을 추천드립니다. 

Q. '여러뷰에서 같은 데이터 값을 활용'하게 된다면 어떤 이점이 있을까요? 
Q. '여러뷰에서 같은 데이터 값을 활용'하지 않고, 각각의 ViewModel로 분리했을 때 어떤 이점이 있을까요? 
Q. MVVM은 어떤 문제를 해결해주기 위한 방법일까요? 
Q. MVVM을 적용하지 않고 구현할 수는 없을까요?
Q. MVVM을 적용한다면 지금 어떤 부분에서 그 의도를 담지 못하고 있는 걸까요?

리팩토링을 본격적으로 시작하셨고, 이후에도 유지보수하기 위해서 스스로 위 질문에 대해 생각해보시고 공부하시다보면 많은 배움이 있을 것 같습니다.
처음부터 바로 적용하기 어렵다면, MVVM이 적용된 코드들을 보시다보면 감을 잡으실 수 있을 것 같습니다 :)

![image](https://github.com/user-attachments/assets/4fffea77-81a9-4d44-9687-737c59891709)


추가로 궁금한 부분이 있으시면 코드와 함께 올려주시면 좋을 것 같습니다 😄

**Reply by SSSiye** [reply]
생각해보지 못한 부분이 많았네요! MVVM이 왜 필요한지부터 생각해보고 적용해보겠습니다. 감사합니다☺️

**Comment by Jaehwa-Noh** [comment]
안녕하세요. ViewModel이 꼭 하나의 View당 대응되는 모델 하나씩 있을 필요는 없습니다.

>그렇다면 제가 사용한 ShoppibngViewModel,ListViewModel과 같은 '여러 뷰에서 같은 데이터 값을 사용하는 것'들도 각각의 ViewModel을 만들어서 관리해야하나요?

이 부분은 추가적인 설명이 필요할 것 같습니다.
ViewModel을 파라미터로 넘긴다는 말씀인가요? 아니면 각 뷰에서 새로운 인스턴스를 만든다는 말씀이신가요?

**Reply by SSSiye** [reply]
지금은 파라미터로 넘기고 있는 상태입니다.
이 방식을 굳이 바꿀 필요는 없는 것이겠지요??

**Reply by Jaehwa-Noh** [reply]
생성하는 곳에서 StateObject, 받는 곳에서 ObservedObject로 받으면 되긴 하는데, ViewModel을 다 넘기기 보다는 필요한 상태(State)를 넘기는 것이 불필요한 화면 갱신을 방지할 수 있습니다.
